pyatlas/pyatlas/spatial_index.py

```diff
@@ -0,0 +1,184 @@
+        raise NotImplementedError('quadtree currently not implemented')
+
+    def add(self, entity):
+        """
+        Insert an AtlasEntity into the index.
+        """
+        if self.tree is None:
+            raise ValueError('must select a tree implementation before using')
+
+        if entity.bounds() is not None:
+            self.tree.add(entity)
+        else:
+            raise ValueError('bounds cannot be None')
+
+    def get(self, bounds, predicate=lambda e: True):
+        """
+        Get a frozenset of AtlasEntities that are within or intersecting some bounds.
+        Can optionally accept a matching predicate function.
+        """
+        if self.tree is None:
+            raise ValueError('must select a tree implementation before using')
+
+        results = []
+        for item_index in self.tree.get(bounds):
+            result = self.atlas.entity(item_index, self.entity_type)
+            if predicate(result):
+                results.append(result)
+
+        return frozenset(results)
+
+
+class _RTree(object):
+    """
+    A wrapper class for the _CustomSTRtree implementation, which calls into the
+    GEOS C library using machinery from Shapely. _CustomSTRtree is immutable,
+    but this class simulates mutability by maintaining a parallel list of
+    elements, and rebuilding the _CustomSTRtree on each add. For this reason,
+    extensive use of the add() method is not recommended.
+
+    Note also, this class stores raw AtlasEntity identifiers without any
+    EntityType information. For this reason, users of _RTree should avoid adding
+    AtlasEntities with different EntityTypes in the same tree.
+    """
+
+    def __init__(self, initial_entities=None):
+        """
+        Create a new _RTree, optionally with an iterable of initial AtlasEntities.
+        """
+        self.contents = []
+        self.tree = None
+
+        if initial_entities is not None:
+            for entity in initial_entities:
+                self.contents.append(entity)
+            self._construct_tree_from_contents()
+
+    def _construct_tree_from_contents(self):
+        """
+        Use the tree's contents list (of AtlasEntities) to rebuild the backing
+        _CustomSTRtree.
+        """
+        contents_shapely_format = [
+            geometry.boundable_to_shapely_box(entity) for entity in self.contents
+        ]
+
+        # pack the arguments in format expected by the _CustomSTRtree
+        hacktree_arguments = []
+        for entity, cont in zip(self.contents, contents_shapely_format):
+            hacktree_arguments.append((entity.get_identifier(), cont))
+
+        self.tree = _CustomSTRtree(hacktree_arguments)
+
+    def add(self, entity):
+        """
+        Insert an AtlasEntity into the _RTree. The underlying STRtree is
+        immutable, so this method forces a rebuild of the entire tree.
+        """
+        self.contents.append(entity)
+        self._construct_tree_from_contents()
```

Comment:
Would it make sense to have two methods, one called "add" and another one called "finishedAdding"? That way you do not build the rtree at every call?

Reply:
I suppose I could change the name to something different, just to make it more clear.

The reason I implemented it this way is because the STR packed R-tree provided by GEOS (thru Shapely) is effectively immutable once it has been built (you can see what I mean by "built" by [checking out this class](https://github.com/OSGeo/geos/blob/master/src/index/strtree/AbstractSTRtree.cpp), "building" occurs as a side effect of queries to the tree ).
So effectively, the only way to add elements to the tree once it has been "built" is to rebuild the tree from scratch. However, this limitation *does not* apply when constructing the tree for the first time, which is why I provided the `initial_entities` parameter in the constructor of the `RTree` class.

Ideally, users will not be building their own `RTrees` anyway. The `Atlas` lazily constructs the spatial indices it needs at query time. However, I made this class is public in case someone really wants to extend the spatial capabilities of an Atlas without necessary modifying the `Atlas` class itself.

Reply:
It would be awesome to add all this as a comment in that method!

Reply:
Haha sounds good, I'll clean this up to be a bit more concise and to the point.